src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -2512,4 +2432,132 @@ std::string primitive_inst::get_implementation_name() const {
+        OPENVINO_ASSERT(node != nullptr);
+        for (auto& impl_manager : m_available_impls) {
+            if ((impl_manager->get_shape_type() & shape_type) != shape_type)
+                continue;
+
+            if (!impl_manager->support_shapes(params))
+                continue;
+
+            return impl_manager->create(*node, params);
+        }
+
+        return nullptr;
+    };
+
+    const auto node = &inst.get_node();
+    auto& prog = *inst.get_network().get_program();
+    auto& kernels_cache = prog.get_kernels_cache();
+
+    // Update param if fake_alignment is available
+    auto updated_params = inst.get_fake_aligned_params_if_possible(params);
+    // Change weights layout of `updated_params` to original one to have valid information
+    // in _impl->_weights_reorder_params about required weights format after impl selection
+    if (inst.get_node().is_type<fully_connected>() || inst.get_node().is_type<convolution>() || inst.get_node().is_type<deconvolution>()) {
+        const auto weights_idx = inst.get_node().get_primitive()->input.size();
+        const auto original_weights_memory = inst.dep_memory_ptr(weights_idx);
+        updated_params.weights_layout = optional_layout(original_weights_memory->get_layout());
+    }
+
+    for (auto& i : updated_params.input_layouts) {
+        i.data_padding._dynamic_dims_mask = padding::EMPTY_MASK;
+    }
+    for (auto& o : updated_params.output_layouts) {
+        o.data_padding._dynamic_dims_mask = padding::EMPTY_MASK;
+    }
+
+    // 1. If we have static impl in the cache - use it
+    if (use_async_compilation && inst.get_impl() && inst.get_impl()->is_dynamic()) {
+        auto cached_impl = m_static_impls_cache.get(updated_params);
+        if (cached_impl) {
+            return cached_impl->clone();
+        }
+
+        // 1.1. Static impl not found - run async compilation
+        auto& compilation_context = prog.get_compilation_context();
+        compilation_context.push_task(updated_params, [&inst, &compilation_context, updated_params, find_impl]() {
+            if (compilation_context.is_stopped())
+                return;
+            auto& _program = *inst.get_network().get_program();
+            auto& cache = _program.get_implementations_cache();
+            {
+                // Check existense in the cache one more time as several iterations of model execution could happens and multiple compilation
+                // tasks created for same shapes
+                if (cache.has(updated_params))
+                    return;
+            }
+
+            std::unique_ptr<primitive_impl> impl = find_impl(&inst.get_node(), updated_params, shape_types::static_shape);
+
+            if (impl->get_kernels_source().size() > 0) {
+                auto kernels = _program.get_kernels_cache().compile(updated_params, impl->get_kernels_source());
+                impl->set_kernels(kernels);
+            }
+            cache.add(updated_params, impl->clone());
```

Comment:
Do we need to clone impl here? Looks like we can use `std::move` here

Reply:
Fixed